File: java/user-management-db-adapter/src/main/java/com/hexagonalchess/usermanagement/InMemoryFriendshipRepository.java
package com.hexagonalchess.usermanagement;

import java.util.ArrayList;
import java.util.List;

public class InMemoryFriendshipRepository implements FriendshipRepository {

    private final List<Friendship> store = new ArrayList<>();

    @Override
    public void save(Friendship friendship) {
        store.add(friendship);
    }

    @Override
    public List<Friendship> findByOwnerId(UserId ownerId) {
        return store.stream()
            .filter(f -> f.ownerId().equals(ownerId))
            .toList();
    }

    @Override
    public void deleteByOwnerAndFriend(UserId ownerId, UserId friendId) {
        store.removeIf(f -> f.ownerId().equals(ownerId) && f.friendId().equals(friendId));
    }
}

File: python/user_management_db_adapter/in_memory_friendship_repository.py
from user_management_policy.friendship import Friendship
from user_management_policy.friendship_repository import FriendshipRepository
from user_management_policy.user_id import UserId


class InMemoryFriendshipRepository(FriendshipRepository):
    def __init__(self):
        self._store: list[Friendship] = []

    def save(self, friendship: Friendship) -> None:
        self._store.append(friendship)

    def find_by_owner_id(self, owner_id: UserId) -> list[Friendship]:
        return [f for f in self._store if f.owner_id == owner_id]

    def delete_by_owner_and_friend(self, owner_id: UserId, friend_id: UserId) -> None:
        self._store = [
            f for f in self._store
            if not (f.owner_id == owner_id and f.friend_id == friend_id)
        ]
